bin/check-dns-zone.rb

```diff
@@ -0,0 +1,193 @@
+#   example commands
+#
+# NOTES:
+#   Does it behave differently on specific platforms, specific use cases, etc
+#
+# LICENSE:
+#   Zubov Yuri <yury.zubau@gmail.com> sponsored by Actility, https://www.actility.com
+#   Released under the same terms as Sensu (the MIT license); see LICENSE
+#   for details.
+#
+
+require 'sensu-plugin/check/cli'
+require 'dnsruby'
+require 'whois'
+require 'whois/parser'
+require 'ipaddr'
+require 'resolv'
+
+#
+# DNS
+#
+class DNSZone < Sensu::Plugin::Check::CLI
+  option :domain,
+         description: 'Domain to resolve (or ip if type PTR)',
+         short: '-d DOMAIN',
+         long: '--domain DOMAIN'
+
+  option :threshold,
+         description: 'Percentage of DNS queries that must succeed',
+         short: '-l PERCENT',
+         long: '--threshold PERCENT',
+         proc: proc(&:to_i),
+         default: 100
+
+  option :timeout,
+         description: 'Set timeout for query',
+         short: '-T TIMEOUT',
+         long: '--timeout TIMEOUT',
+         proc: proc(&:to_i),
+         default: 5
+
+  option :warn_only,
+         description: 'Warn instead of critical on failure',
+         short: '-w',
+         long: '--warn-only',
+         boolean: true
+
+  def resolve_ns
+    Resolv::DNS.new.getresources(config[:domain], Resolv::DNS::Resource::IN::NS).map { |e| e.name.to_s }
+  end
+
+  def check_whois(entries)
+    errors = []
+    success = []
+    record = Whois.whois(config[:domain])
+    parser = record.parser
+    additional_text = "(whois #{parser.nameservers.map(&:name)}) (dig #{entries})"
+    if Set.new(parser.nameservers.map(&:name)) == Set.new(entries)
+      success << "Resolved #{config[:domain]} #{config[:type]} equal with whois #{additional_text}"
+    else
+      errors << "Resolved #{config[:domain]} #{config[:type]} did not include #{config[:result]} #{additional_text}"
+    end
+    [errors, success]
+  end
+
+  def check_results(entries)
+    errors = []
+    success = []
+
+    %w[check_whois check_axfr soa_query].each do |check|
+      result = send(check, entries)
+      errors.push(*result[0])
+      success.push(*result[1])
+    end
+
+    result = check_dns_connection(entries, true)
+    errors.push(*result[0])
+    success.push(*result[1])
+
+    result = check_dns_connection(entries, false)
+    errors.push(*result[0])
+    success.push(*result[1])
+
+    [errors, success]
+  end
+
+  def check_dns_connection(entries, use_tcp = true)
+    errors = []
+    success = []
+
+    entries.each do |ns|
+      Dnsruby::Resolv.getaddresses(ns).each do |ip|
+        begin
+          resolv = Dnsruby::Resolver.new(nameserver: ip.to_s, do_caching: false, use_tcp: use_tcp)
+          resolv.query_timeout = config[:timeout]
+          resolv.query(config[:domain], Dnsruby::Types.ANY)
+
+          success << "Resolved DNS #{ns}(#{ip}) uses #{use_tcp ? 'tcp' : 'udp'}"
+        rescue StandardError
+          errors << "Resolved DNS #{ns}(#{ip}) doesn't use #{use_tcp ? 'tcp' : 'udp'}"
```

Comment:
rather than using the same conditional again just assign it once out of the rescue block and use it for both inside the begin and rescue.